Sources/SourceKitLSP/Swift/CodeActions/ApplyDeMorganLaw.swift

```diff
@@ -0,0 +1,730 @@
+        let only = tuple.elements.only
+      {
+        switch Self(expr: only.expression) {
+        case .single(_, let desinglified), .other(let desinglified):
+          self = .single(tuple, desinglified: desinglified)
+        }
+      } else {
+        self = .other(expr)
+      }
+    }
+
+    /// Negates the expression of interest, maps the negated expression with the single ancestor if any, and returns
+    /// the result.
+    ///
+    /// - Parameters:
+    ///   - exprType: The type of expression that this and all recursive negations must agree with.
+    ///
+    /// - Returns: The orignal expression negated, and the type of the outermost change taken place.
+    func negated(exprType: ExprType) -> Negated {
+      if let complementOfNegation = self.negatedNegation(exprType: exprType) {
+        return complementOfNegation
+      }
+      if let complementOfPropositions = self.negatedPropositions(exprType: exprType) {
+        return complementOfPropositions
+      }
+      if let complementOfComparison = self.negatedComparison(exprType: exprType) {
+        return complementOfComparison
+      }
+      if let complementOfTernary = self.negatedTernary(exprType: exprType) {
+        return complementOfTernary
+      }
+      return self.map(negation: Negation(exprType: exprType, expr: self.exprOfInterest))
+    }
+
+    /// If the expression of interest is a negation expression, returns the negated expression and the type of change
+    /// as `Change.denegation`.
+    ///
+    /// The only valid form of a negation expression is `NOT Expr`,
+    /// ```swift
+    /// // boolean
+    /// !a, !true
+    /// !(a && b), !(a || b)
+    /// !(a is Int), !(b as! Bool), !(a >= b), !(a ? true : false)
+    ///
+    /// // bitwise
+    /// ~b, ~0
+    /// ~(a | b), ~(a & b)
+    /// ~(a as Int), ~(b << 1), ~(a ? 1 : 0)
+    /// ```
+    ///
+    /// - Note: A valid negation expression must have its prefix agree with `exprType`.
+    func negatedNegation(exprType: ExprType) -> Negated? {
+      if let prefixExpr = self.as(PrefixOperatorExprSyntax.self),
+        let negatedNegation = Negation(prefixExpr, exprType: exprType)
+      {
+        // TODO: remove parentheses?
+        (self.map(negatedExpr: negatedNegation.expr), .denegation)
+      } else {
+        nil
+      }
+    }
+
+    /// If the expression of interest is propositions, returns its negated expression and the type of change as either
+    /// `Change.orToAnd` or `Change.andToOr`.
+    ///
+    /// - Note: A valid instance of propositions must have its binary operator agree with `exprType`.
+    func negatedPropositions(exprType: ExprType) -> Negated? {
+      if let infixExpr = self.as(InfixOperatorExprSyntax.self),
+        let negatedPropositions = NegatedPropositions(infixExpr, exprType: exprType)
+      {
+        (
+          self.map(negatedExpr: negatedPropositions.expr),
+          negatedPropositions.operator.kind == .and ? .orToAnd : .andToOr
```

Comment:
Can you add explicit `return` here and in the `else` clause because swift-syntax needs to compile with Swift 5.8, which doesn’t have `if` expressions. I think the same applies to a few other methods as well.

Reply:
Isn't SourceKitLSP compiled with Swift 5.9?

Reply:
You are right. The code just uses so many swift-syntax types that I thought I was in the swift-syntax repo.

I still think that explicit returns would help readability here, though. Especially with the returned tuple.